### PR TITLE
Moderation Abstract classes and implementations

### PR DESCRIPTION
**Pull request by DeveloperCallum**
Added abstract classes to streamline the checking. its untested because the bot refuses to start for me.

Testing - I have not testing the code, they were small changes and im too lazy to set everything up. If you see any logical errors or errors when CI/CD testing just @ me before pulling. (I assume you will test it anyway)

### CODE REVIEW

**Comment by DynxstyGIT** [comment]
Looks good so far, please fix all the Checkstyle problems tho

**Comment by DeveloperCallum** [comment]
> 

No idea what you mean by that, ive never seen this before. is it just a format check?

**Comment by MoonTM-GIT** [comment]
Please test the bot before marking this as ready for review again. *I've told him how to set up the bot*

**Comment by DynxstyGIT** [comment]
@DeveloperCallum

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import net.dv8tion.jda.api.entities.Member;
+import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
+import net.dv8tion.jda.api.requests.restaction.interactions.ReplyCallbackAction;
+import net.javadiscord.javabot.command.ResponseException;
+import net.javadiscord.javabot.command.Responses;
+import net.javadiscord.javabot.command.interfaces.ISlashCommand;
+
+/**
+ * Basic moderation command.
+ */
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:13]:
```suggestion
/**
 * Abstract class, that represents a single moderation command.
 */
```

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public abstract class ModerateCommand implements ISlashCommand {
+	private boolean allowThreads = true;
+
+	@Override
+	public ReplyCallbackAction handleSlashCommandInteraction(SlashCommandInteractionEvent event) throws ResponseException {
+		if (event.getGuild() == null) {
+			return Responses.error(event, "Moderation commands can only be used inside guilds.");
+		}
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:21]:
```suggestion
		if (event.getGuild() == null) {
			return Responses.error(event, "This command can only be used inside servers.");
		}
```

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateUserCommand.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+		if (target == null) {
+			return Responses.error(event, "Cannot report a user who is not a member of this server");
+		}
+
+		if (!actOnSelf || !commandUser.isOwner()) {
+			if (commandUser.getId().equals(target.getId())) {
+				return Responses.error(event, "You cannot preform actions on yourself.");
+			}
+		}
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateUserCommand.java:47]:
```suggestion
		if (!actOnSelf || !commandUser.isOwner()) {
			if (commandUser.getId().equals(target.getId())) {
				return Responses.error(event, "You cannot perform this action on yourself.");
			}
		}
```

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateUserCommand.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		if (target.isOwner()) {
+			return Responses.error(event, "You cannot preform actions on a higher member staff member.");
+		}
+
+		//If both users have at least one role.
+		if (target.getRoles().size() > 0 && commandUser.getRoles().size() > 0) {
+			if (commandUser.getRoles().get(0).getPosition() > target.getRoles().get(0).getPosition()) {
+				return Responses.error(event, "You cannot preform actions on a higher member staff member.");
+			}
+		}
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateUserCommand.java:58]: I think you could just check if `Member#canInteract(Member)` is true

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+		Member member = event.getMember();
+
+		if (member == null) {
+			return Responses.error(event, "Unexpected error has occurred, Slash Command member was null.");
+		}
+
+		return handleModerationCommand(event, member);
+	}
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:40]:
```suggestion
		Member member = event.getMember();
		if (member == null) {
			return Responses.error(event, "Unexpected error has occurred, Slash Command member was null.");
		}
		return handleModerationCommand(event, member);
	}
```

**Review by DynxstyGIT — CHANGES_REQUESTED** [review]

**Comment by DynxstyGIT** [comment]
And make sure to test every command affected by this, if not done already

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java (hunk 5 above)
Comment (DeveloperCallum) [reply src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:40]: No idea, why this was moved/changed.

Comment (DynxstyGIT) [reply src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:40]: Just removed some spaces to match with the rest of the code

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java (hunk 2 above)
Comment (DynxstyGIT) [reply src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:21]: I think we should still stick with how discord officially calls servers/guilds
Guild is just an internal term and may cause confusion

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java (hunk 5 above)
Comment (DynxstyGIT) [reply src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:40]: Btw, you can just accept/decline this suggestion here on GitHub. You don't need to make a separate commit in intellij

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java (hunk 2 above)
Comment (MoonTM-GIT) [reply src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:21]: I agree.

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+			}
+		} else {
+			if (event.getChannelType() != ChannelType.TEXT) {
+				return Responses.error(event, "This command can only be performed in a server text channel.");
+			}
+		}
+		Member member = event.getMember();
+		if (member == null) {
+			return Responses.error(event, "Unexpected error has occurred, Slash Command member was null.");
+		}
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:35]: Make this error message a bit more user friendly

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateUserCommand.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+		if (targetOption == null) {
+			return Responses.error(event, "Missing required arguments.");
+		}
+
+		target = targetOption.getAsMember();
+
+		if (target == null) {
+			return Responses.error(event, "Cannot report a user who is not a member of this server");
+		}
```
Comment (DynxstyGIT) [inline src/main/java/net/javadiscord/javabot/command/moderation/ModerateUserCommand.java:41]:
```suggestion
		OptionMapping targetOption = event.getOption("user");
		if (targetOption == null) {
			return Responses.error(event, "Missing required arguments.");
		}
        Member target = targetOption.getAsMember();
		if (target == null) {
			return Responses.error(event, "Cannot report a user who is not a member of this server");
		}
```

**Review by DynxstyGIT — CHANGES_REQUESTED** [review]

**Comment by DeveloperCallum** [comment]
Yea later, it was fine before I commited your requested changed. idk why it suddently has decided the format is wrong. because nothing but words changed.

File: src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java (hunk 6 above)
Comment (DeveloperCallum) [reply src/main/java/net/javadiscord/javabot/command/moderation/ModerateCommand.java:35]: There are 4 comments there, which...

There are as friendly as they can be, and it should never reach `member == null`, unless the event was emitted.

**Review by DynxstyGIT — APPROVED** [review]

**Pull request merged**